File: backend/app/schemas/execution.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.domain.enums.execution_status import ExecutionStatus
from app.domain.enums.language import Language


class ExecuteRequest(BaseModel):
    file_id: uuid.UUID


class ExecutionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    file_id: uuid.UUID
    user_id: uuid.UUID
    language: Language
    status: ExecutionStatus
    stdout: str | None
    stderr: str | None
    exit_code: int | None
    duration_ms: int | None
    created_at: datetime
    started_at: datetime | None
    completed_at: datetime | None


class ExecutionSummary(BaseModel):
    """Lightweight shape for execution history listings — omits stdout/stderr."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    file_id: uuid.UUID
    language: Language
    status: ExecutionStatus
    exit_code: int | None
    duration_ms: int | None
    created_at: datetime
    started_at: datetime | None
    completed_at: datetime | None


class ExecutionWithProject(ExecutionSummary):
    project_name: str


class ExecutionStats(BaseModel):
    total: int
    successful: int
    failed: int
    last_activity_at: datetime | None

File: frontend/types/execution.ts
export type ExecutionStatus = "queued" | "running" | "success" | "failed" | "timeout" | "cancelled";

export interface Execution {
  id: string;
  project_id: string;
  file_id: string;
  user_id: string;
  language: string;
  status: ExecutionStatus;
  stdout: string | null;
  stderr: string | null;
  exit_code: number | null;
  duration_ms: number | null;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
}

export interface ExecutionSummary {
  id: string;
  project_id: string;
  file_id: string;
  language: string;
  status: ExecutionStatus;
  exit_code: number | null;
  duration_ms: number | null;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
}

export const TERMINAL_STATUSES: ExecutionStatus[] = ["success", "failed", "timeout", "cancelled"];
